### PR TITLE
Added (Local) Matrix of futures of tiles.

### PR DESCRIPTION
**Pull request by rasolca**
- Added local Matrix of futures of tiles and its tests.
- Fixed LayoutInfo which was failing for some corner cases.

### CODE REVIEW

File: test/unit/test_matrix.cpp [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+      set(mat, el);
+
+      CHECK_MATRIX_EQ(el, mat);
+    }
+  }
+}
+
+template <class T, Device device>
+void checkFromExisting(T* p, const matrix::LayoutInfo& layout, Matrix<T, device>& matrix) {
+  auto el = [](const GlobalElementIndex& index) {
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: Please, use better names for `el` and `el2`, e.g. `reference_matrix1`, `reference_matrix2`.

File: test/unit/test_matrix.cpp [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+  EXPECT_EQ(MatrixBase(layout), matrix);
+  CHECK_MATRIX_PTR(ptr, matrix);
+  CHECK_MATRIX_EQ(el, matrix);
+
+  set(matrix, el2);
+
+  for (SizeType j = 0; j < size.cols(); ++j) {
+    for (SizeType i = 0; i < size.rows(); ++i) {
+      if (el2({i, j}) != *ptr({i, j})) {
+        FAIL() << "Error at index (" << i << ", " << j << "): "
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: In general, you can use every GoogleTest macro as a stream (see [doc](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions))

In this particular case, `ASSERT_EQ(...) << "additional message"`

File: test/unit/test_matrix.cpp [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+    const Type* p = mem();
+    Matrix<const Type, Device::CPU> mat(layout, p, mem.size());
+
+    CHECK_FROM_EXISTING(mem(), layout, mat);
+  }
+}
+
+/// @brief Returns true if the first @p ready futures are ready.
+/// @pre Future* should be future or shared_future
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: The * stands for 1 or 2...maybe something similar to FutureN?

File: test/unit/test_matrix.cpp [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    memory::MemoryView<Type, Device::CPU> mem(layout.minMemSize());
+
+    const Type* p = mem();
+    Matrix<const Type, Device::CPU> mat(layout, p, mem.size());
+
+    CHECK_FROM_EXISTING(mem(), layout, mat);
+  }
+}
+
+/// @brief Returns true if the first @p ready futures are ready.
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: I would say _"Return turn if JUST the firsts [...]"_ and try to find a more explicit name for the function.

File: test/unit/test_matrix.cpp [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+      return false;
+  }
+  return true;
+}
+
+/// @brief Checks if current[i] depends correctly on previous[i].
+/// If get_ready == true it checks if current[i] is ready after previous[i] is used.
+/// If get_ready == false it checks if current[i] is not ready after previous[i] is used.
+/// @pre Future* should be future or shared_future
+template <template <class> class Future1, template <class> class Future2, class T1, class T2,
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: You are testing futures and you are not really interested in the data itself.

I would simplify this with
```cpp
template <class Future1, class Future2>
void checkFutures(bool get_ready, const std::vector<Future1>& current, std::vector<Future2>& previous)
```

I would apply this also to the other functions whose concern is just the future (e.g. `checkFuturesStep`).

File: test/unit/test_matrix.cpp [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+          Device device>
+void checkFutures(bool get_ready, const std::vector<Future1<Tile<T1, device>>>& current,
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: Can we find a better name? In this function, you are not just checking `current`, but you are also triggering `previous`. I would make it more explicit.

IMHO it would make more clear also the role of `get_ready`, it could become something like `current_ready_after_trigger`.

File: test/unit/test_matrix.cpp [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+  EXPECT_TRUE(checkFuturesStep(get_ready ? current.size() : 0, current));
+}
+
+TYPED_TEST(MatrixTest, Dependencies) {
+  using Type = TypeParam;
+
+  for (const auto& size : sizes) {
+    for (const auto& block_size : block_sizes) {
+      Matrix<Type, Device::CPU> mat(size, block_size);
+
```
Comment (albestro) [inline test/unit/test_matrix.cpp]: I would add a comment that shows the pipeline you are testing.

```bash
             shfut2a
            /       \
fut0 - fut1          fut3 - shfut4a
            \       /
             shfut2b


             shfut2a       shfut4a
            /       \     /       \
fut0 - fut1          fut3          fut5
            \       /     \       /
             shfut2b       shfut4b
```

**Comment by albestro** [comment]
About my concerns about Matrix/Tile API.

I may agree on the "reference-semantic" of these objects (i.e. Tile, Matrix), but I would investigate better if there is a way to avoid this overload and the non-const-from-const inheritance "workaround".

After the discussion, we agreed that, in case, we can try refactoring this in the next early stages of the development.

File: src/matrix/layout_info.cpp [hunk 8]
```diff
@@ -44,18 +45,20 @@ LayoutInfo::LayoutInfo(const GlobalElementSize& size, const TileElementSize& blo
...
+    if (ld_tile_ < tmp) {
       throw std::invalid_argument("Error: Invalid Leading Dimension");
     }
-    if (tile_offset_row_ < static_cast<std::size_t>(block_size_.rows())) {
+    if (tile_offset_row_ < static_cast<std::size_t>(tmp)) {
       throw std::invalid_argument("Error: Invalid Tile Row Offset");
     }
-    if (tile_offset_row_ < minTileSize(block_size_) &&
+    if (tile_offset_row_ < minTileMemSize(block_size_) &&
         static_cast<std::size_t>(ld_tile_) < sum(tileOffset({nr_tiles_.rows() - 1, 0}), last_rows)) {
```
[the remark this answers is not in the document]
Comment (rasolca) [reply src/matrix/layout_info.cpp]: Debatable about readability.
Anyway you are reviewing code that didn't change.

File: include/dlaf/matrix.h [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                                       T* ptr, std::size_t elements) {
+  return Matrix<T, device>(matrix::tileLayout(size, block_size), ptr, elements);
+}
+
+template <class T, Device device>
+Matrix<T, device> createMatrixFromTile(const GlobalElementSize& size, const TileElementSize& block_size,
+                                       SizeType ld_tile, SizeType tiles_per_col, T* ptr,
+                                       std::size_t elements) {
+  return Matrix<T, device>(matrix::tileLayout(size, block_size, ld_tile, tiles_per_col), ptr, elements);
+}
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.h]: When the matrix will be distributed, the construction will not be so straightforward.

File: include/dlaf/matrix.h [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+class Matrix : public MatrixBase {
+public:
+  using ElementType = T;
+  using TileType = Tile<ElementType, device>;
+  using ConstTileType = Tile<const ElementType, device>;
+  friend Matrix<const ElementType, device>;
+
+  Matrix(const GlobalElementSize& size, const TileElementSize& block_size);
+
+  Matrix(const matrix::LayoutInfo& layout, ElementType* ptr, std::size_t elements);
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.h]: I thought this was clear (first comment of the PR):
> Documentation for methods which involve a layout description will be added when the Matrix will support distributed matrices as well.

File: include/dlaf/matrix.ipp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  return tile_shared_futures_[i];
+}
+
+template <class T, Device device>
+hpx::future<Tile<T, device>> Matrix<T, device>::operator()(const LocalTileIndex& index) {
+  std::size_t i = tileLinearIndex(index);
+  hpx::future<TileType> old_future = std::move(tile_futures_[i]);
+  hpx::promise<TileType> p;
+  tile_futures_[i] = p.get_future();
+  tile_shared_futures_[i] = {};
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.ipp]: I don't understand.
`shared_future` does not have a `clear()` method.

File: include/dlaf/matrix.h [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+#include <vector>
+#include "dlaf/matrix/layout_info.h"
+#include "dlaf/matrix_base.h"
+#include "dlaf/tile.h"
+#include "dlaf/types.h"
+
+namespace dlaf {
+
+template <class T, Device device>
+class Matrix : public MatrixBase {
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.h]: const correctness do not always respect the DRY principle (e.g. vector operator[])
Moreover `tileLinearIndex` in an implementation detail, and, in general, the order in which the tiles are stored is indifferent.
regarding setUpTiles, not only the Tile type changes but also the future type. IMHO setUpTiles do not contradict the DRY principle.

File: test/include/dlaf_test/util_matrix.h [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+/// err_message(el({i, j}), (i, j)-element) is printed for the first element which does not fulfill the comparison.
+/// @pre el argument is an index of type const GlobalElementIndex&.
+/// @pre comp should have two arguments and return true if the comparison is fulfilled and false otherwise.
+/// @pre err_message should have two arguments and return a string.
+/// @pre el return type should be the same as the type of the first argument of comp adn of err_message.
+/// @pre The second argument of comp should be either T, T& or const T&.
+/// @pre The second argument of err_message should be either T, T& or const T&.
+namespace internal {
+template <class T, class Func1, class Func2, class Func3>
+void check(Matrix<T, Device::CPU>& mat, Func1 el, Func2 comp, Func3 err_message, const char* file,
```
[the remark this answers is not in the document]
Comment (rasolca) [reply test/include/dlaf_test/util_matrix.h]: This function do not assert. As written in the doc it checks if the elements fulfill the comparison and add a failure when it finds the first element that doesn't.
The function and macros which re supposed to be used by users (this is an internal detail) have the self explaining name.

File: test/unit/test_matrix.cpp (hunk 6 above)
Comment (rasolca) [reply test/unit/test_matrix.cpp]: Ok that the name should explain what the function does, but here we are going in the direction that the function name should contain the full documentation.

File: include/dlaf/matrix.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.h]: > As per the discussion, I'm still not convinced of this API, see review comment for additional details.
We cannot start a discussion on the API on each PR. We will never progress.
The "pointer" semantic is the only way for us to provide all the functionalities we need, therefore we use it.

File: test/unit/test_matrix.cpp (hunk 3 above)
Comment (rasolca) [reply test/unit/test_matrix.cpp]: FutureN is even more confusing IMO.

File: include/dlaf/matrix.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix.h]: 🙏 Sorry, I forgot it.

What about accepting just the `MemoryView&&` and not the raw pointer?

File: include/dlaf/matrix.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix.h]: Ok, I respect your point.

What about inverting the template parameters?

File: test/unit/test_matrix.cpp (hunk 1 above)
Comment (rasolca) [reply test/unit/test_matrix.cpp]: `reference_matrix` - > I would expect a matrix.
This is how `el` was chosen:
`matrix_elements` ~> `elements` ~>`el`
Is `elements` ok?

Comment (albestro) [reply test/unit/test_matrix.cpp]: It was just my opinion. I agree with you that `elements` does not change so much from `el`.

File: include/dlaf/matrix.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix.h]: Will do it

File: test/include/dlaf_test/util_matrix.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (albestro) [reply test/include/dlaf_test/util_matrix.h]: I said it was an assert-ish thing because, looking at the code I see, as you said, that you want to stop at the first failed comparison. I would have made it clear from the call that this is going to assert and stop the test.

However, it is ok for me. 👌🏼

File: include/dlaf/matrix.h [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+  /// TODO: Sync details.
+  /// @pre index.isValid() == true.
+  /// @pre index.isIn(nrTiles()) == true.
+  hpx::shared_future<ConstTileType> read(const LocalTileIndex& index) noexcept;
+
+protected:
+  /// @brief Returns the position in the vector of the index Tile.
+  /// @pre index.isValid() == true.
+  /// @pre index.isIn(nrTiles()) == true.
+  std::size_t tileLinearIndex(const LocalTileIndex& index) noexcept {
```
Comment (albestro) [inline include/dlaf/matrix.h]: `const`?

File: include/dlaf/matrix.ipp [hunk 15]
```diff
@@ -62,21 +48,8 @@ hpx::future<Tile<T, device>> Matrix<T, device>::operator()(const LocalTileIndex&
 template <class T, Device device>
 void Matrix<T, device>::setUpTiles(const memory::MemoryView<T, device>& mem,
                                    const matrix::LayoutInfo& layout) {
-  using util::size_t::mul;
-  const auto& nr_tiles = layout.nrTiles();
-  ld_futures_ = static_cast<std::size_t>(nr_tiles.rows());
-  std::size_t tot_tiles = mul(ld_futures_, nr_tiles.cols());
-  tile_futures_.reserve(tot_tiles);
-  tile_shared_futures_.resize(tot_tiles);
+  ld_futures_ = static_cast<std::size_t>(layout.nrTiles().rows());
```
Comment (albestro) [inline include/dlaf/matrix.ipp]: I would initialize `ld_futures_` in the constructors.

File: include/dlaf/matrix.h [hunk 16]
```diff
@@ -102,28 +97,44 @@ class Matrix<const T, device> : public MatrixBase {
...
+
+  void setUpConstTiles(const memory::MemoryView<ElementType, device>& mem,
+                       const matrix::LayoutInfo& layout);
 
+  template <template <class> class Future, class TileT>
+  void setUpTilesInternal(std::vector<Future<TileT>>& tile_futures_vector,
+                          const memory::MemoryView<ElementType, device>& mem,
+                          const matrix::LayoutInfo& layout);
+
+  std::size_t futureVectorSize(const matrix::LayoutInfo& layout);
```
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix.h]: I would also add `noexcept`.

**Comment by rasolca** [comment]
- Added local Matrix of futures of tiles and its tests.
- Fixed LayoutInfo which was failing for some corner cases.
